File: src/connection.rs
use std::io;

use crate::response::Result;
use crate::response::ResponseStatus;
use crate::response::FileStat;
use crate::socket::Socket;
use crate::requests::RequestType;

use tracing::info;

pub struct Connection {
    socket: Socket,
}

const SOCKET_HANDSHAKE: u32 = 0xFAFE5ABE;

impl Connection {
    pub fn connect(address: &str, port: u16) -> io::Result<Self>
    {
        let socket = Socket::connect(address, port)?;
        
        let mut connection = Connection {
            socket
        };
        connection.handshake()?;
        
        Ok(connection)
    }

    fn handshake(&mut self) -> io::Result<()>
    {
        self.socket.write_u32(SOCKET_HANDSHAKE)?;
        self.socket.flush()?;

        self.socket.read_u32()
            .and_then(|hs| {
                if hs != SOCKET_HANDSHAKE {
                    return Err(io::Error::new(io::ErrorKind::InvalidData, 
                        format!("Recieved handshake did not match expected value! (expected {:X}, recieved {:X})", SOCKET_HANDSHAKE, hs)
                    ));
                }

                info!("Handshake succeeded!");
                Ok(())
            })
    }

    pub fn req_getattr(&mut self, path: &str) -> Result<FileStat> {
        self.socket.write_byte(RequestType::StatFile.into())?;
        self.socket.write_string(path)?;
        self.socket.flush()?;

        match ResponseStatus::from(self.socket.read_byte()?) {
            ResponseStatus::Success => {},
            other => return Err(other),
        };

        Ok(
            FileStat {
                mode: self.socket.read_u16()?,
                size: self.socket.read_u64()?,
                access_time: self.socket.read_u64()?,
                write_time: self.socket.read_u64()?,
            }
        )
    }
}

File: src/connection.cpp
#include "connection.hpp"
#include "logger.hpp"

#include <format>

static const int BUFFER_SIZE = 2048;
static const uint32_t HANDSHAKE_DATA = 0xFAFE5ABE;

namespace nandroid
{
    Connection::Connection(const std::string& address, uint16_t port): socket(address, port), reader(&socket, BUFFER_SIZE), writer(&socket, BUFFER_SIZE)
    {
        perform_handshake();
    }

    void Connection::perform_handshake()
    {
        writer.write_u32(HANDSHAKE_DATA);
        writer.flush();

        uint32_t received_data = reader.read_u32();
        if(received_data != HANDSHAKE_DATA)
        {
            throw std::runtime_error(std::format("Handshake failed (expected {:X}, received {:X})", HANDSHAKE_DATA, received_data));
        }
    }

    ResponseStatus Connection::req_readdir(const std::string& path, const std::function<void(const std::string&, const nandroidfs::FileStat&)>& consume)
    {
        try
        {
            writer.write_byte((uint8_t)nandroidfs::RequestType::ListDirectory);
            writer.write_utf8_string(path);
            writer.flush();
    
            ResponseStatus status = (ResponseStatus)reader.read_byte();
            if(status != ResponseStatus::Success)
            {
                return status;
            }
    
            while((status = (ResponseStatus)reader.read_byte()) != ResponseStatus::NoMoreEntries)
            {
                if(status == ResponseStatus::Success)
                {
                    std::string filename = reader.read_utf8_string();
                    FileStat stat(reader);
    
                    consume(filename, stat);
                }
            }
        }
        catch(const std::exception& e)
        {
            nandroid::Logger::error("Caught exception!\t{}", e.what());
            return ResponseStatus::GenericFailure;
        }

        return ResponseStatus::Success;
    }

    ResponseStatus Connection::req_getattr(const std::string& path, FileStat& out)
    {
        try 
        {
            writer.write_byte((uint8_t)RequestType::StatFile);
            writer.write_utf8_string(path);
            writer.flush();
    
            ResponseStatus status = (ResponseStatus)reader.read_byte();
            if(status != ResponseStatus::Success)
            {
                return status;
            }
    
            out = FileStat(reader);
        }
        catch(const std::exception& e)
        {
            nandroid::Logger::error("Caught exception!\t{}", e.what());
            return ResponseStatus::GenericFailure;
        }

        return ResponseStatus::Success;
    }

    ResponseStatus Connection::req_open(const std::string& path, nandroidfs::OpenMode mode, bool read, bool write, int& out)
    {
        try
        {
            writer.write_byte((uint8_t)RequestType::OpenHandle);
            nandroidfs::OpenHandleArgs args(path, mode, read, write);
            args.write(writer);
            writer.flush();

            ResponseStatus status = (ResponseStatus)reader.read_byte();
            if(status != ResponseStatus::Success)
            {
                return status;
            }

            out = reader.read_u32();
        }
        catch(const std::exception& e)
        {
            nandroid::Logger::error("Caught exception!\t{}", e.what());
            return ResponseStatus::GenericFailure;
        }

        return ResponseStatus::Success;
    }

    ResponseStatus Connection::req_release(int handle)
    {
        try
        {
            writer.write_byte((uint8_t)RequestType::CloseHandle);
            writer.write_u32(handle);
            writer.flush();

            return (ResponseStatus)reader.read_byte();
        }
        catch(const std::exception& e)
        {
            nandroid::Logger::error("Caught exception!\t{}", e.what());
            return ResponseStatus::GenericFailure;
        }
    }

    ResponseStatus Connection::req_read(int handle, uint8_t* buffer, size_t size, off_t offset, int& bytes_read)
    {
        try
        {
            writer.write_byte((uint8_t)RequestType::ReadHandle);
            nandroidfs::ReadHandleArgs args(handle, size, offset);
            args.write(writer);
            writer.flush();
    
            ResponseStatus status = (ResponseStatus)reader.read_byte();
            if(status != ResponseStatus::Success)
            {
                return status;
            }
    
            int read =reader.read_u32();
            reader.read_exact(buffer, read);
            bytes_read = read;
        }
        catch(const std::exception& e)
        {
            nandroid::Logger::error("Caught exception!\t{}", e.what());
            return ResponseStatus::GenericFailure;
        }

        return ResponseStatus::Success;
    }
}